keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1330 @@
+- `.status.nodes[0].PodsEvacuating`, to show how many pods have been accepted for the
+  evacuation from the first node. These are the pods that have the `EvacuationInitiated`
+  condition set to `True`.
+- To keep track of the entire maintenance the controller will reconcile a `Drained` condition and
+  set it to true if all pods pending evacuation/termination have terminated on all target nodes
+  when drain is requested by the maintenance object.
+- NodeMaintenance condition or annotation can be set on the node object to advertise the current
+  phase of the maintenance.
+
+#### Supported Stage Transitions
+
+The following transitions should be validated by the API server.
+
+- Idle -> _Deletion_
+  - Planning a maintenance in the future and canceling/deleting it without any consequence.
+- (Idle) -> Cordon -> (Complete) -> _Deletion_.
+  - Make a set of nodes unschedulable and then schedulable again.
+  - The complete stage will always be run even without specifying it.
+- (Idle) -> (Cordon) -> Drain -> (Complete) -> _Deletion_.
+  - Make a set of nodes unschedulable, drain them, and then make them schedulable again.
+  - Cordon and Complete stages will always be run, even without specifying them.
+- (Idle) -> Complete -> _Deletion_.
+  -  Make a set of nodes schedulable.
+
+### Deployment and ReplicaSet Controllers
+
+The replica set controller will watch its pods and count the number of pods it observes with a
+`EvacuationRequested` condition. It will then store this count in `.status.ReplicasToEvacuate`
+
+The deployment controller will watch its ReplicaSets and react when it observes positive number of
+pods in `.status.ReplicasToEvacuate`. If the owning object of the targeted pods is a Deployment with
+a positive `.spec.strategy.rollingUpdate.maxSurge` value, the controller will create surge pods by
+scaling up the ReplicaSet. The new pods will not be scheduled on the maintained node because the
+`.spec.unschedulable` field would be set to true on that node. As soon as the surge pods become
+available, the deployment controller will scale down the ReplicaSet. The replica set controller
+will then in turn delete the pods with the `EvacuationRequested` condition.
+
+For completeness, the deployment controller will also track the total number of targeted pods of
+all its ReplicaSets under its `.status.ReplicasToEvacuate`.
+
+If the node maintenance prematurely ends before the surge process has a chance to complete, the
+deployment controller will scale down the ReplicaSet which will then remove the extra pods that
+were created during the surge.
+
+```golang
+type ReplicaSetStatus struct {
+    ...
+    ReplicasToEvacuate int32
+    ...
+}
+```
+
+```golang
+type DeploymentStatus struct {
+    ...
+    ReplicasToEvacuate int32
+    ...
+}
+```
+
+To support providing a response to the drain process that the evacuation has begun. The deployment
+controller will annotate all replica sets that support the evacuation with the annotation
```

Comment:
What does it mean `to support the evacution`? And why use an annotation and not a proper spec field?

Reply:
The evacuation process has been changed. PTAL at https://github.com/kubernetes/enhancements/pull/4565